common/quotas/cluster_aware_quota_calculator.go

```diff
@@ -0,0 +1,69 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+type (
+	// MemberCounter returns the total number of instances there are for a given service.
+	MemberCounter interface {
+		MemberCount() int
+	}
+	// ClusterAwareQuotaCalculator calculates the available quota for the current host based on the per instance and per
+	// cluster quota. The quota could represent requests per second, total number of active requests, etc. It works by
+	// dividing the per cluster quota by the total number of instances running the same service.
+	ClusterAwareQuotaCalculator quotaCalculator[func() int]
+	// ClusterAwareNamespaceSpecificQuotaCalculator is similar to ClusterAwareQuotaCalculator, but it uses quotas that
+	// are specific to a namespace.
+	ClusterAwareNamespaceSpecificQuotaCalculator quotaCalculator[func(namespace string) int]
+	// quotaCalculator is a generic type that we use because the quota functions could be namespace specific or not.
+	quotaCalculator[T any] struct {
+		MemberCounter MemberCounter
```

Comment:
you could make this a `func() int` and then it'd be easier to fake (and just pull out `.MemberCount` for the real one)? I like it this way too

Reply:
I started doing this, but I ran into 2 problems:
1. There's stil a nil check here, and I'm not sure that it's safe to remove, and I don't want to do something like serviceResolver.MemberCount and have that trigger an NPE somewhere 
2. I think that because `ServiceResolver` implements this interface, it's clearer what should be passed in here vs if it was just a `func() int`.